cs130_back/handlers/user.go

```diff
@@ -489,6 +489,120 @@ func GetClassesInfo(db *gorm.DB, w http.ResponseWriter, r *http.Request) {
 	respondWithJSON(w, http.StatusOK, response)
 }
 
+// UserSearchDetails contains details displayed during user search
+type UserSearchDetails struct {
+	ID				int
+	FirstName 		string		
+	LastName		string
+	Email			string
+
+}
+
+// GetAllUsers retrieves all users in the DB (for search functionality)
+func GetAllUsers(db *gorm.DB, w http.ResponseWriter, r *http.Request) {
+	response := make(map[string][]UserSearchDetails)
+
+	rows, err := db.Model(&models.User{}).Select("id, first_name, last_name, email").Rows()
+	if err != nil {
+		return
+	}
+	defer rows.Close()
+
+	var UserDetails []UserSearchDetails
+	for rows.Next() {
+		var details UserSearchDetails
+		db.ScanRows(rows, &details)
+		UserDetails = append(UserDetails, details)
+	}
+
+	response["users"] = UserDetails
+	respondWithJSON(w, http.StatusOK, response)
+}
+
+// CourseViewDetails contains all the information needed for a Course page
+type CourseViewDetails struct {
+	CourseName		string
+	StudyBuddies 	[]StudyBuddy
+	Listings 		[]models.Listing
+}
+
+// StudyBuddy is a User object without certain 'private' fields
+type StudyBuddy struct {
+	ID				int		`json:"u_id"`
+	FirstName		string	`json:"first_name"`
+	LastName		string	`json:"last_name"`
+	Email			string	`json:"u_email"`
+	Biography		string	`json:"biography"`
+	Timezone		string 	`json:"timezone"`
+	SchoolName		string 	`json:"school_name"` 
+	Availability	pq.Int64Array	`json:"availability"`
+}
+
+// GetBuddiesAndListings retrieves Study Buddies and Listings for each of a user's courses
+func GetBuddiesAndListings(db *gorm.DB, w http.ResponseWriter, r *http.Request) {
+	vars := r.URL.Query()
+	id, ok := strconv.Atoi(vars["u_id"][0])
+	if ok != nil {
+		respondWithError(w, http.StatusBadRequest, "Invalid user id")
+		return
+	}
+
+	//Get user object
+	p := models.User{ID: id}
+	if GetUserByID(db, &p, w) == 0 {
+		return
+	}
+
+	//Get all courses
+	var courses []models.Course
+	err := p.GetCourses(db, &courses)
+	if err != nil {
+		return
+	}
+
+	//returns map of courseIDs to their CourseViewDetails
+	response := make(map[int]CourseViewDetails)
+
+	for _, j := range courses {
+		var tempDetails CourseViewDetails
+		tempDetails.CourseName = j.Name
```

Comment:
We need the courseId too for AddListing later. Otherwise, everything else looks good to me.

Reply:
The way it currently is, it returns a map of CourseIDs to their CourseViewDetails. Is it not convenient to use this way?

Reply:
oh my b, didn't notice that, lgtm then.